### PR TITLE
Fixes #33694 - upgrade prettier to v2

### PR DESCRIPTION
**Pull request by sharvit**
Just upgrading prettier to v2 and running `npm run lint -- --fix`

- It is possible to reduce the diff by playing with the settings, for example setting `arrowParens` to `avoid` but I'm not sure if it is worth it, opinions?
  https://prettier.io/docs/en/options.html#arrow-function-parentheses

### CODE REVIEW

File: package.json [hunk 1]
```diff
@@ -49,7 +49,7 @@
     "highlight.js": "~9.14.0",
     "node-sass": "^4.5.0",
     "optimize-css-assets-webpack-plugin": "^3.2.0",
-    "prettier": "^1.19.1",
+    "prettier": "2.4.1",
```
Comment (sharvit) [inline package.json:52]: Seems like prettier recommend using `exact-same` version: https://prettier.io/docs/en/install.html

File: script/npm_install_plugins.js [hunk 2]
```diff
@@ -2,7 +2,7 @@
 const childProcess = require('child_process');
 const { packageJsonDirs } = require('./plugin_webpack_directories');
 
-packageJsonDirs('pipe').forEach(pluginPath => {
+packageJsonDirs('pipe').forEach((pluginPath) => {
```
Comment (tbrisker) [inline script/npm_install_plugins.js:5]: I'm not sure about this rule, looks like our current style is the opposite and it leads to quite a lot of changes. On the other hand I personally find it a bit more readable.

File: script/plugin_webpack_directories.js [hunk 3]
```diff
@@ -6,20 +6,17 @@ var fs = require('fs');
 
 // If we get multiple lines, then the plugin_webpack_directories.rb script
 // has on the stdout more that just the JSON we want, so we use newline to split and check.
-var sanitizeWebpackDirs = pluginDirs => {
-  var splitDirs = pluginDirs
-    .toString()
-    .split('\n')
-    .reverse();
+var sanitizeWebpackDirs = (pluginDirs) => {
+  var splitDirs = pluginDirs.toString().split('\n').reverse();
```
Comment (tbrisker) [inline script/plugin_webpack_directories.js:10]: 👍🏼 to longer lines before requiring a break (if this is the rule that changed here)

File: webpack/assets/javascripts/foreman_hosts.js [hunk 4]
```diff
@@ -45,8 +44,9 @@ export function getAttributesToPost(componentType) {
 }
 
 export function copyRegistrationCommand() {
-  const commandText = document.getElementById('registration_command')
-    .textContent;
+  const commandText = document.getElementById(
+    'registration_command'
+  ).textContent;
```
Comment (tbrisker) [inline webpack/assets/javascripts/foreman_hosts.js:49]: 👎🏼 to the rule that led to this change, looks much worse imho

File: webpack/assets/javascripts/react_app/common/withReactRoutes.js [hunk 5]
```diff
@@ -4,12 +4,13 @@ import { Router } from 'react-router-dom';
...
 
-const withReactRoutes = Component => props => (
-  <Router history={history}>
-    <AppSwitcher>
-      <Component {...props} />
-    </AppSwitcher>
-  </Router>
-);
+const withReactRoutes = (Component) => (props) =>
+  (
```
Comment (tbrisker) [inline webpack/assets/javascripts/react_app/common/withReactRoutes.js:8]: not sure about this rule either, i think the previous version might be more readable?

File: webpack/assets/javascripts/react_app/components/AuditsList/ExpansiveView.js [hunk 6]
```diff
@@ -80,12 +80,8 @@ class ExpansiveView extends React.Component {
   }
 
   render() {
-    const {
-      comment,
-      actionDisplayName,
-      auditedChangesWithIdToLabel,
-      details,
-    } = this.props;
+    const { comment, actionDisplayName, auditedChangesWithIdToLabel, details } =
```
Comment (tbrisker) [inline webpack/assets/javascripts/react_app/components/AuditsList/ExpansiveView.js:83]: This change is worse I think, it makes looking at the history more difficult if you want to figure out when a certain prop was added and why

**Review by tbrisker — CHANGES_REQUESTED** [review]
I tried to look over the changes and think which ones are better and which aren't and commented inline on an example of each rule (i think). Would it be possible to upgrade the version and disable all the new rules first (or at least the ones causing most changes), and then open another PR for each rule to discuss separately if we want to implement it or not? That is how we did it with rubocop rules for the ruby code.

**Comment by ekohl** [comment]
Also a general note that we've also been cautious with Rubocop massive refactors. They make cherry picks harder and often patches need to be rebased to resolve conflicts.

File: script/npm_install_plugins.js (hunk 2 above)
Comment (sharvit) [reply script/npm_install_plugins.js:5]: Yes, they changed the default value for https://prettier.io/docs/en/options.html#arrow-function-parentheses

**Comment by sharvit** [comment]
> I tried to look over the changes and think which ones are better and which aren't and commented inline on an example of each rule (i think). Would it be possible to upgrade the version and disable all the new rules first (or at least the ones causing most changes), and then open another PR for each rule to discuss separately if we want to implement it or not? That is how we did it with rubocop rules for the ruby code.

The issue is, prettier is not customizable by design/philosophy because they believe in forcing the community opinion so you don't need to waste your time on discussions:

> “We want to free mental threads and end discussions around style. While sometimes fruitful, these discussions are for the most part wasteful.”
> "Even if Prettier does not format all code 100% the way you’d like, it’s worth the “sacrifice” given the unique benefits of Prettier"
- https://prettier.io/docs/en/option-philosophy.html
- https://prettier.io/docs/en/why-prettier.html

tbh I prefer the way the code looks with v1 but I don’t think we should use a project/version which is not maintained for the long term.
On the other side, I do like the approach of using whatever ever works and is agreed by the wider community so we don't need to think or discuss that.

**Comment by ekohl** [comment]
https://github.com/theforeman/foreman/pull/8839#discussion_r730376166 is an example where a cherry pick is made harder because of coding style. I wonder if it would be better to disable prettier in stable branches just to avoid that. It wouldn't really solve things because many would no longer cleanly apply, but it'd reduce the pain a bit.

**Comment by ares** [comment]
We decided to switch this to draft as the broader conversation on discourse is necessary first.

**Comment by sharvit** [comment]
> We decided to switch this to draft as the broader conversation on discourse is necessary first.

Sounds good, I would like to bring this discussion again after the branching

**Comment by adamruzicka** [comment]
Has the discussion already happened? Does anyone have a link to it?

**Comment by ekohl** [comment]
AFAIK it hasn't.

**Comment by ezr-ondrej** [comment]
Thanks @sharvit I believe this is opened for too long and won't get rebased.
I'm closing for now, but if @Ron-Lavi or @MariaAga feel like it, it would be worth revisiting ^_^

**Pull request closed without merging**